feat: add mozilla/sccache.

monaka opened the pull request:
Signed-off-by: Masaki Muranaka <monaka@monami-ya.com>

---

[mozilla/sccache](https://github.com/mozilla/sccache): sccache is ccache with cloud storage

Installation: https://github.com/mozilla/sccache#installation

> There are prebuilt x86-64 binaries available for Windows, Linux (a portable binary compiled against musl), and macOS [on the releases page](https://github.com/mozilla/sccache/releases/latest)

monaka committed:
feat: add mozilla/sccache.

Signed-off-by: Masaki Muranaka <monaka@monami-ya.com>

suzuki-shunsuke commented:
Thank you for your contribution!

suzuki-shunsuke commented:
Please run `aqua-registry scaffold --deep` next time.

https://aquaproj.github.io/docs/products/aqua-registry/contributing/#how-to-add-packages

```console
$ aqua-registry scaffold --deep mozilla/sccache
```

This command outputs many error messages in case of this package, then please fix the configuration to resolve errors.

suzuki-shunsuke commented:
Let me fix.

monaka commented:
@suzuki-shunsuke I tried to run `aqua-registry scaffold --deep mozilla/sccache`.
It removes my registry.yaml. 
Logs are like this.

```
+ aqua gr --deep --out-testdata pkgs/mozilla/sccache/pkg.yaml mozilla/sccache > pkgs/mozilla/sccache/registry.yaml
WARN[0000] get the repository                            aqua_version=2.8.0 env=linux/amd64 error="GET https://api.github.com/repos/mozilla/sccache: 403 API rate limit exceeded for 49.105.12.26. (But here's the good news: Authenticated requests get a higher rate limit. Check out the documentation for more details.) [rate reset in 34m10s]" program=aqua repo_name=sccache repo_owner=mozilla
WARN[0000] list releases                                 aqua_version=2.8.0 env=linux/amd64 error="GET https://api.github.com/repos/mozilla/sccache/releases?per_page=100: 403 API rate limit of 60 still exceeded until 2023-06-14 12:40:19 +0900 JST, not making remote request. [rate reset in 34m10s]" program=aqua repo_name=sccache repo_owner=mozilla
Update registry.yaml
appending '- import: pkgs/mozilla/sccache/pkg.yaml' to aqua-local.yaml
+ aqua i --test
ERRO[0000] install the package                           aqua_version=2.8.0 env=linux/amd64 error="invalid package: github_release package requires asset" package_name=mozilla/sccache package_version=v0.5.3 program=aqua registry=standard
ERRO[0000] install the package                           aqua_version=2.8.0 env=linux/amd64 error="invalid package: github_release package requires asset" package_name=mozilla/sccache package_version=v0.5.3 program=aqua registry=standard
ERRO[0000] install the package                           aqua_version=2.8.0 env=linux/amd64 error="invalid package: github_release package requires asset" package_name=mozilla/sccache package_version=v0.5.3 program=aqua registry=standard
ERRO[0000] install the package                           aqua_version=2.8.0 env=linux/amd64 error="invalid package: github_release package requires asset" package_name=mozilla/sccache package_version=v0.5.3 program=aqua registry=standard
ERRO[0000] install the package                           aqua_version=2.8.0 env=linux/amd64 error="invalid package: github_release package requires asset" package_name=mozilla/sccache package_version=v0.5.3 program=aqua registry=standard
ERRO[0000] install the package                           aqua_version=2.8.0 env=linux/amd64 error="invalid package: github_release package requires asset" package_name=mozilla/sccache package_version=v0.5.3 program=aqua registry=standard
FATA[0000] aqua failed                                   aqua_version=2.8.0 env=linux/amd64 error="it failed to install some packages" program=aqua
FATA[0000] aqua failed                                   aqua_version=0.1.10 env=linux/amd64 error="execute a command: aqua i: exit status 1" program=aqua-registry
```

```
~/git/aqua-registry$ cat pkgs/mozilla/sccache/registry.yaml 
packages:
  - type: github_release
    repo_owner: mozilla
    repo_name: sccache
```

suzuki-shunsuke commented:
```
403 API rate limit exceeded
```

This is due to API rate limit.

suzuki-shunsuke commented:
And before running `aqua-registry scaffold --deep mozilla/sccache`, please clean up `pkgs/mozilla/sccache`.

```console
$ rm -Rf pkgs/mozilla/sccache
```

Otherwise, I guess `pkgs/mozilla/sccache/pkg.yaml` will be wrong.

suzuki-shunsuke commented:
After running `aqua-registry scaffold --deep mozilla/sccache`, please set `files`.

```console
$ vi pkgs/mozilla/sccache/registry.yaml
$ aqua-registry gr
```

```yaml
    files:
      - name: sccache
        src: 'sccache-{{.Version}}-{{.Arch}}-{{.OS}}/sccache'
```

suzuki-shunsuke committed:
fix(mozilla/sccache): aqua-registry scaffold --deep mozilla/sccache

```console
$ rm -Rf pkgs/mozilla/sccache
$ aqua-registry scaffold --deep mozilla/sccache
```

suzuki-shunsuke committed:
fix(mozilla/sccache): set files

suzuki-shunsuke committed:
fix(mozilla/sccache): disable checksum on windows

suzuki-shunsuke commented:
v4.18.0 is out 🎉
https://github.com/aquaproj/aqua-registry/releases/tag/v4.18.0

monaka commented:
Thank you for your additional patches and merging.

I've noticed a lack of `files:`.
And I couldn't figure out why `aqua-registy scaffold --deep mozilla/sccache` removes it and shows many errors. This is a reason why I couldn't finish to create my patch.

It looks that my situation can be reproduced on the main branch.
Should I publish a new issue?

```
$ aqua-registry version
aqua-registry version 0.1.10 (15b76d73554236ec5242122ca12312f85fdf9634)
```

```
git checkout main
aqua-registy scaffold --deep mozilla/sccache
```
So many errors... https://gist.github.com/monaka/ba1f17804aba45b0f2f4d33f6a0f74d8

And the result:

```
$ git diff --no-pager pkgs/
$ git --no-pager diff pkgs
diff --git a/pkgs/mozilla/sccache/registry.yaml b/pkgs/mozilla/sccache/registry.yaml
index d4bd24fab..68511ddae 100644
--- a/pkgs/mozilla/sccache/registry.yaml
+++ b/pkgs/mozilla/sccache/registry.yaml
@@ -5,9 +5,6 @@ packages:
     description: sccache is ccache with cloud storage
     asset: sccache-{{.Version}}-{{.Arch}}-{{.OS}}.{{.Format}}
     format: tar.gz
-    files:
-      - name: sccache
-        src: sccache-{{.Version}}-{{.Arch}}-{{.OS}}/sccache
     overrides:
       - goos: windows
         replacements:
@@ -56,14 +53,6 @@ packages:
           - darwin
           - amd64
       - version_constraint: semver(">= 0.2.14")
-        overrides:
-          - goos: windows
-            replacements:
-              arm64: arm64
-            checksum:
-              # https://github.com/aquaproj/aqua-registry/pull/13027
-              # https://github.com/aquaproj/aqua-registry/actions/runs/5262947082/jobs/9512626767
-              enabled: false # ERRO[0001] install the package                           actual_checksum=BA6BF184D2B0A78978629AAEF0D9FE3CA57923E228D949B1932EC1A91AB4CEAC aqua_version=2.8.0 env=windows/amd64 error="checksum is invalid" expected_checksum="��B\x00A\x006\x00B\x00F\x001\x008\x004\x00D\x002\x00B\x000\x00A\x007\x008\x009\x007\x008\x006\x002\x009\x00A\x00A\x00E\x00F\x000\x00D\x009\x00F\x00E\x003\x00C\x00A\x005\x007\x009\x002\x003\x00E\x002\x002\x008\x00D\x009\x004\x009\x00B\x001\x009\x003\x002\x00E\x00C\x001\x00A\x009\x001\x00A\x00B\x004\x00C\x00E\x00A\x00C\x00" package_name=mozilla/sccache package_version=0.2.14 program=aqua registry=standard
       - version_constraint: semver(">= 0.2.13")
         overrides: []
         replacements:
@@ -75,12 +64,7 @@ packages:
           - darwin
         rosetta2: true
       - version_constraint: semver(">= 0.2.9")
-        overrides:
-          - goos: windows
-            checksum:
-              # https://github.com/aquaproj/aqua-registry/pull/13027
-              # https://github.com/aquaproj/aqua-registry/actions/runs/5262947082/jobs/9512626767
-              enabled: false # ERRO[0001] install the package                           actual_checksum=6DD85F65FCCB1CDB00E7C5804C3F3BA523C1ED20EA8F777D454826EEF17C636F aqua_version=2.8.0 env=windows/amd64 error="checksum is invalid" expected_checksum="��6\x00D\x00D\x008\x005\x00F\x006\x005\x00F\x00C\x00C\x00B\x001\x00C\x00D\x00B\x000\x000\x00E\x007\x00C\x005\x008\x000\x004\x00C\x003\x00F\x003\x00B\x00A\x005\x002\x003\x00C\x001\x00E\x00D\x002\x000\x00E\x00A\x008\x00F\x007\x007\x007\x00D\x004\x005\x004\x008\x002\x006\x00E\x00E\x00F\x001\x007\x00C\x006\x003\x006\x00F\x00" package_name=mozilla/sccache package_version=0.2.9 program=aqua registry=standard
+        overrides: []
         replacements:
           amd64: x86_64
           darwin: apple-darwin
```

monaka commented:
I tried using `rm -Rf pkgs/mozilla/sccache/`. The result was same.

suzuki-shunsuke commented:
`aqua-registy scaffold --deep mozilla/sccache` overwrites `pkgs/mozilla/sccache/registry.yaml`.
This command doesn't support patching.
And as I mentioned this command doesn't generate `files` https://github.com/aquaproj/aqua/issues/1667 .

About the comment ddc8c2cbee04782972b7e3f282e2e34ca81ee41c , I added the change to avoid the following error.

```
ERRO[0001] install the package                           actual_checksum=BA6BF184D2B0A78978629AAEF0D9FE3CA57923E228D949B1932EC1A91AB4CEAC aqua_version=2.8.0 env=windows/amd64 error="checksum is invalid" expected_checksum="��B\x00A\x006\x00B\x00F\x001\x008\x004\x00D\x002\x00B\x000\x00A\x007\x008\x009\x007\x008\x006\x002\x009\x00A\x00A\x00E\x00F\x000\x00D\x009\x00F\x00E\x003\x00C\x00A\x005\x007\x009\x002\x003\x00E\x002\x002\x008\x00D\x009\x004\x009\x00B\x001\x009\x003\x002\x00E\x00C\x001\x00A\x009\x001\x00A\x00B\x004\x00C\x00E\x00A\x00C\x00" package_name=mozilla/sccache package_version=0.2.14 program=aqua registry=standard
```

I'm not sure why the checksum verification failed.
I haven't seen the error. The expected checksum looks weird. Maybe this is a bug of aqua, but I'm not sure.
I'd like to investigate this when I have time, but the priority is low.

suzuki-shunsuke commented:
> The expected checksum looks weird

Maybe the checksum file of https://github.com/mozilla/sccache is broken. I'll take a look.

suzuki-shunsuke commented:
Oh, it seems the character encoding is not UTF-8 but UTF-16.

```console
$ cat sccache-0.2.14-x86_64-pc-windows-msvc.tar.gz.sha256 
??BA6BF184D2B0A78978629AAEF0D9FE3CA57923E228D949B1932EC1A91AB4CEAC%

$ cat sccache-0.2.9-x86_64-pc-windows-msvc.tar.gz.sha256
??6DD85F65FCCB1CDB00E7C5804C3F3BA523C1ED20EA8F777D454826EEF17C636F%
```

```console
$ nkf -g sccache-0.2.9-x86_64-pc-windows-msvc.tar.gz.sha256
UTF-16

$ nkf -g sccache-0.2.14-x86_64-pc-windows-msvc.tar.gz.sha256
UTF-16
```
